jupyter_client/connect.py

```diff
@@ -24,12 +26,17 @@
 from .localinterfaces import localhost
 from .utils import _filefind
 
+if TYPE_CHECKING:
+    from jupyter_client import BlockingKernelClient
+
+    from .session import Session
+
 # Define custom type for kernel connection info
 KernelConnectionInfo = Dict[str, Union[int, str, bytes]]
 
 
 def write_connection_file(
-    fname: Optional[str] = None,
+    fname: str | None = None,
```

Comment:
I dont' think we should do this if we still support 3.8 and 3.9, this beciome valid syntax, but `typing.get_type_hints()` will fail on `|` for Union syntax on 3.8 and 3.9, so we risk breaking things.

Reply:
```
from __future__ import annotations
from typing import Optional, get_type_hints


def write_connection_file(fname: str | None = None) -> None:
    pass


get_type_hints(write_connection_file)
```
---
```
$ python f.py
Traceback (most recent call last):
  File "/Users/bussonniermatthias/dev/jupyter_client/f.py", line 9, in <module>
    get_type_hints(write_connection_file)
  File "/Users/bussonniermatthias/miniforge3/envs/39/lib/python3.9/typing.py", line 1386, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/Users/bussonniermatthias/miniforge3/envs/39/lib/python3.9/typing.py", line 254, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "/Users/bussonniermatthias/miniforge3/envs/39/lib/python3.9/typing.py", line 493, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
```

Reply:
Yeah, not your fault, it's a common confusion. I wouldn't have caught it if not for a similar problem recently.

Reply:
Well now I'm not so sure what to do, because this also applies to `type[]`, `set[]`, `list[]`, `dict[]`, etc.  I don't think we need to claim to support *type hints* for Python 3.8 or 3.9.

Reply:
At least not at run time.

Reply:
Nothing prevents folks from calling `get_type_hints` on their own code, I just verified.

Reply:
as you made the PR i think it's fine, but maybe we ca avoid this for subsequent libs.

Reply:
Yeah that sounds like a good plan.  I added this, which I'll use as a template:

```
  # non-pep585-annotation
  "UP006",
  # non-pep604-annotation
  "UP007",
```